### PR TITLE
Add cli driver for all clients

### PR DESCRIPTION
**Pull request by amcp**
Setup on EC2 Amazon Linux:
```bash
sudo yum update -y \
&& sudo yum upgrade -y \
&& sudo yum install -y java-1.8.0-openjdk-devel git emacs \
&& sudo yum remove -y java-1.7.0-openjdk \
&& git clone https://github.com/amcp/ndbench.git \
&& cd ndbench \
&& git checkout cli \
&& ./gradlew build
```

Test 1:
```bash
DISCOVERY_ENV=AWS ./gradlew run \
-Dndbench.config.cli.clientName=DynamoDBKeyValue \
-Dndbench.config.cli.bulkSize=1 \
-Dndbench.config.cli.timeoutMillis=180000 \
-Dndbench.config.dataSize=1000 \
-Dndbench.config.numKeys=1 \
-Dndbench.config.readEnabled=true \
-Dndbench.config.readRateLimit=20000 \
-Dndbench.config.writeEnabled=true \
-Dndbench.config.writeRateLimit=20000 \
-Dndbench.config.dynamodb.rcu=20000 \
-Dndbench.config.dynamodb.wcu=20000 \
-Dndbench.config.dynamodb.tablename=hires \
-Dndbench.config.dynamodb.consistentread=true \
-Dndbench.config.dynamodb.maxRetries=0 \
-Dndbench.config.dynamodb.requestTimeout=100 \
-Dndbench.config.dynamodb.maxConnections=400 \
-Dndbench.config.dynamodb.numReaders=200 \
-Dndbench.config.dynamodb.numWriters=200 \
-Dndbench.config.dynamodb.programmableTables=true
```

### CODE REVIEW

File: ndbench-cli/build.gradle [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+sourceCompatibility = 1.8
+
+repositories {
+    mavenCentral()
+}
+
+dependencies {
+    compile project(':ndbench-core')
+    compile project(':ndbench-api')
+    compile (project(':ndbench-cass-plugins')) {
```
[the remark this answers is not in the document]
Comment (amcp) [reply ndbench-cli/build.gradle]: fixed

File: ndbench-cli/src/main/java/com/netflix/ndbench/core/NdbenchEntrypoint.java [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+ * This class is a CLI entry point to facilitate quick testing of the Netflix Database Benchmark.
+ * In particular, this class does not require deploying a WAR to Tomcat to run the benchmark.
+ * Nor does this class require running a Context in a web container.
+ *
+ * TODO:
+ * - support other clients (databases)
+ * - add command line arguments for all generic and client-specific parameters
+ * -
+ *
+ * @author amcp@amazon.com
```
[the remark this answers is not in the document]
Comment (amcp) [reply ndbench-cli/src/main/java/com/netflix/ndbench/core/NdbenchEntrypoint.java]: fixed

File: ndbench-cli/src/main/java/com/netflix/ndbench/core/NdbenchEntrypoint.java [hunk 3]
```diff
@@ -0,0 +1,346 @@
+/*
+ *  Copyright 2018 Amazon Web Services, Inc.
```
[the remark this answers is not in the document]
Comment (amcp) [reply ndbench-cli/src/main/java/com/netflix/ndbench/core/NdbenchEntrypoint.java]: fixed

File: ndbench-cli/src/main/java/com/netflix/ndbench/core/NdbenchEntrypoint.java [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+            }
+
+            @Override
+            public Integer getMaxConnections() {
+                return args.maxConnections;
+            }
+        });
+
+        try {
+            driver.init(dynamodb); //TODO make database client configurable
```
[the remark this answers is not in the document]
Comment (amcp) [reply ndbench-cli/src/main/java/com/netflix/ndbench/core/NdbenchEntrypoint.java]: not implementing other drivers now, so put this feedback off.

File: ndbench-core/src/main/java/com/netflix/ndbench/core/generators/RandomStringKeyGenerator.java [hunk 5]
```diff
@@ -31,7 +31,7 @@ public RandomStringKeyGenerator(boolean preLoadKeys, int numKeys) {
 
     @Override
     public String getNextKey() {
-        int randomKeyIndex = kRandom.nextInt(getNumKeys());
+        int randomKeyIndex = kRandom.nextInt(numKeys);
```
[the remark this answers is not in the document]
Comment (amcp) [reply ndbench-core/src/main/java/com/netflix/ndbench/core/generators/RandomStringKeyGenerator.java]: #163 fixes #160

File: ndbench-dynamodb-plugins/build.gradle [hunk 6]
```diff
@@ -4,7 +4,6 @@ repositories {
    }
 }
 
-
```
[the remark this answers is not in the document]
Comment (amcp) [reply ndbench-dynamodb-plugins/build.gradle]: fixed

File: ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java [hunk 7]
```diff
@@ -252,8 +252,8 @@ public String writeSingle(String key) throws Exception {
 
     private List<WriteRequest> generateWriteRequests(List<String> keys) {
         return keys.stream()
-                .map(key -> ImmutableMap.of(partitionKeyName, new AttributeValue(key),
-                        ATTRIBUTE_NAME, new AttributeValue(this.dataGenerator.getRandomValue())))
+                .map(key -> ImmutableMap.of("id", new AttributeValue(key),
```
[the remark this answers is not in the document]
Comment (amcp) [reply ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java]: fixed

File: settings.gradle [hunk 8]
```diff
@@ -1,3 +1,13 @@
 rootProject.name='ndbench'
-include 'ndbench-web','ndbench-api','ndbench-core','ndbench-cass-plugins','ndbench-sample-plugins','ndbench-dyno-plugins','ndbench-es-plugins','ndbench-geode-plugins','ndbench-janusgraph-plugins', 'ndbench-evcache-plugins', 'ndbench-dynamodb-plugins'
-
+include 'ndbench-web'
```
[the remark this answers is not in the document]
Comment (amcp) [reply settings.gradle]: fixed #161 in #164

**Comment by amcp** [comment]
evcache causes exceptions related to Eureka on startup. Any ideas?

**Comment by amcp** [comment]
Also, seems like the properties are not getting initialized, despite my customizations:
```
DISCOVERY_ENV=AWS ./gradlew run \
-Dndbench.config.cli.clientName=DynamoDBKeyValue \
-Dndbench.config.cli.bulkSize=25 \
-Dndbench.config.cli.timeoutMillis=360000 \
-Dndbench.config.dynamodb.maxConnections=112 \
-Dndbench.config.dynamodb.numWriters=35 \
-Dndbench.config.dynamodb.numReaders=77 \
-Dndbench.config.dataSize=1000 \
-Dndbench.config.numKeys=10000000 \
-Dndbench.config.readEnabled=true \
-Dndbench.config.readRateLimit=3000 \
-Dndbench.config.writeEnabled=true \
-Dndbench.config.writeRateLimit=3000
```

Result:
```
17:41:49,395  INFO NdbenchEntrypoint:43 - Starting driver in CLI with loadPattern=random, windowSize=-1, windowDurationInSec=-1, bulkSize=1, timeout(ms)=360000, clientName=InMemoryTest
```

**Comment by amcp** [comment]
@ipapapa I fixed the system property issue by modifying build.gradle. The following command now works:

```bash
DISCOVERY_ENV=AWS ./gradlew run \
-Dndbench.config.cli.clientName=DynamoDBKeyValue \
-Dndbench.config.cli.bulkSize=25 \
-Dndbench.config.cli.timeoutMillis=360000 \
-Dndbench.config.dataSize=1000 \
-Dndbench.config.numKeys=100000000 \
-Dndbench.config.readEnabled=true \
-Dndbench.config.readRateLimit=3000 \
-Dndbench.config.writeEnabled=true \
-Dndbench.config.writeRateLimit=3000 \
-Dndbench.config.dynamodb.consistentread=true \
-Dndbench.config.dynamodb.maxRetries=0 \
-Dndbench.config.dynamodb.requestTimeout=100 \
-Dndbench.config.dynamodb.maxConnections=62 \
-Dndbench.config.dynamodb.numWriters=36 \
-Dndbench.config.dynamodb.numReaders=26
```

**Comment by ipapapa** [comment]
In regards to Eureka, I am aware of the issue. It is also reported in #134 (cc @smadappa). I got sidetracked last week, but I am planning to work unless @smadappa beats me to it :)

**Comment by ipapapa** [comment]
LGTM... @vinaykumarchella ?

File: ndbench-es-plugins/src/test/java/com/netflix/ndbench/plugin/es/AutotuneTest.java [hunk 9]
```diff
@@ -15,7 +16,7 @@
     private static final Logger Logger = LoggerFactory.getLogger(AutotuneTest.class);
 
 	
-    @Test
+    @Ignore
```
[the remark this answers is not in the document]
Comment (amcp) [reply ndbench-es-plugins/src/test/java/com/netflix/ndbench/plugin/es/AutotuneTest.java]: fixed

File: ndbench-cli/src/main/java/com/netflix/ndbench/core/NdbenchEntrypoint.java [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+            driver.stop(); //blocking
+            Logger.info("Stopped driver");
+            System.exit(0);
+        } catch(Exception e) {
+            Logger.error("Encountered an exception when driving load", e);
+            System.exit(-1);
+        }
+    }
+
+    static class StaticConfiguration implements IConfiguration {
```
[the remark this answers is not in the document]
Comment (amcp) [reply ndbench-cli/src/main/java/com/netflix/ndbench/core/NdbenchEntrypoint.java]: fixed

File: ndbench-cli/src/main/java/com/netflix/ndbench/core/NdbenchEntrypoint.java [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+ *
+ * TODO:
+ * - support other clients (databases)
+ * - add command line arguments for all generic and client-specific parameters
+ * -
+ *
+ * @author amcp@amazon.com
+ */
+public class NdbenchEntrypoint {
+    public static class Args {
```
[the remark this answers is not in the document]
Comment (amcp) [reply ndbench-cli/src/main/java/com/netflix/ndbench/core/NdbenchEntrypoint.java]: fixed

File: ndbench-core/src/main/java/com/netflix/ndbench/core/generators/StringKeyGenerator.java [hunk 12]
```diff
@@ -38,13 +40,7 @@ protected StringKeyGenerator(int numKeys, boolean preloadKeys) {
...
     public void init() {
         if (this.isPreLoadKeys()) {
-            logger.info("Preloading " + numKeys + " keys");
-            for (int i = 0; i < getNumKeys(); i++) {
-                if (i % 10000 == 0)
-                    logger.info("Still initializing sample data for Keys. So far: "+ i+" /"+numKeys);
-                keys.add("T" + i);
-            }
-            logger.info("Preloaded " + numKeys + " keys");
+            keys.addAll(IntStream.of(getNumKeys()).boxed().map(i -> "T" + i).collect(Collectors.toList()));
```
Comment (amcp) [inline ndbench-core/src/main/java/com/netflix/ndbench/core/generators/StringKeyGenerator.java]: add a peek for the info statement

**Comment by vinaykumarchella** [comment]
Just back from short vacation, will go through it today and provide my comments on this

File: ndbench-cli/src/main/java/com/netflix/ndbench/cli/NdbenchCli.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+import com.google.inject.Injector;
+import com.netflix.ndbench.cli.config.CliConfigs;
+import com.netflix.ndbench.core.NdBenchClientFactory;
+import com.netflix.ndbench.core.NdBenchDriver;
+import com.netflix.ndbench.core.config.GuiceInjectorProvider;
+import com.netflix.ndbench.core.util.LoadPattern;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is a CLI entry point to facilitate quick testing of the Netflix Database Benchmark.
```
[the remark this answers is not in the document]
Comment (amcp) [reply ndbench-cli/src/main/java/com/netflix/ndbench/cli/NdbenchCli.java]: fixed

File: ndbench-cli/src/main/java/com/netflix/ndbench/cli/NdbenchCli.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+                    + ", timeout(ms)=" + cliConfigs.getCliTimeoutMillis()
+                    + ", clientName=" + cliConfigs.getClientName());
+            driver.start(
+                    LoadPattern.fromString(cliConfigs.getLoadPattern()),
+                    Integer.valueOf(cliConfigs.getWindowSize()),
+                    Integer.valueOf(cliConfigs.getWindowDurationInSec()),
+                    Integer.valueOf(cliConfigs.getBulkSize())
+            );
+            long millisToWait = Integer.valueOf(cliConfigs.getCliTimeoutMillis());
+            logger.info("Waiting " + millisToWait + " ms for reads and writes to finish");
```
[the remark this answers is not in the document]
Comment (amcp) [reply ndbench-cli/src/main/java/com/netflix/ndbench/cli/NdbenchCli.java]: With programmatic tables and large amounts of IOPS, it is easy to forget to delete the table after a Control-C and incur extra cost. The timeout is an extra safeguard to control the cost of experiments. I agree about overloading the meaning of non-positive numbers. Will fix.

**Pull request merged**